File: 9-18.py
INF = int(1e8)
E = [[INF]*204 for _ in range(204)]
D = [[0]*204 for _ in range(204)]

def solution(n, s, a, b, fares):
    for i in range(1, n + 1, 1):
        E[i][i] = 0

    for u, v, w in fares:
        E[u][v] = E[v][u] = w

    for i in range(1, n + 1, 1):
        for j in range(1, n + 1, 1):
            D[i][j] = E[i][j]
    for k in range(1, n + 1): # k: 거쳐가는 정점
        for i in range(1, n + 1): # i: 출발 정점
            for j in range(1, n + 1): # j: 도착 정점
                if D[i][k] + D[k][j] < D[i][j]:
                    D[i][j] = D[i][k] + D[k][j]

    answer = INF
    for k in range(1, n + 1, 1):
        ret = D[s][k] + D[k][a] + D[k][b]
        answer = min(answer, ret)
        
    return answer

File: 9-18.cpp
#include <bits/stdc++.h>
using namespace std;

int E[204][204], D[204][204];
const int INF = (int)1e8;

int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int i, j, k;
    for (i = 1; i <= n; i++)
        for (j = 1; j <= n; j++)
            if (i != j) E[i][j] = INF;

    for (int i = 0; i < fares.size(); i++) {
        int u = fares[i][0], v = fares[i][1], w = fares[i][2];
        E[u][v] = E[v][u] = w;
    }

    for (i = 1; i <= n; i++)
        for (j = 1; j <= n; j++)
            D[i][j] = E[i][j];
    for (k = 1; k <= n; k++)
        for (i = 1; i <= n; i++)
            for (j = 1; j <= n; j++)
                if (D[i][k] + D[k][j] < D[i][j])
                    D[i][j] = D[i][k] + D[k][j];

    int answer = D[s][a] + D[s][b];
    for (k = 1; k <= n; k++) {
        if (s == k) continue;
        int ret = D[s][k] + D[k][a] + D[k][b];
        answer = min(answer, ret);
    }
    return answer;
}